firedrake/solving_utils.py

```diff
@@ -214,9 +217,14 @@ def form_function(self, snes, X, F):
 
         assemble(ctx.Fs[lvl], tensor=ctx._Fs[lvl],
                  form_compiler_parameters=problem.form_compiler_parameters)
-        # no mat_type -- it's a vector!
-        for bc in problem.bcs:
-            bc.zero(ctx._Fs[lvl])
+
+        if self.pre_apply_bcs:
```

Comment:
FWIW, I fix this here https://github.com/firedrakeproject/firedrake/blob/better-solvers/firedrake/solving_utils.py#L269 (branch still WIP)